File: samples/java/GameQuest/Shared/src/main/java/systems/zlink/samples/gamequest/shared/contracts/Messages.java
package systems.zlink.samples.gamequest.shared.contracts;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.List;

public final class Messages {
    private Messages() {}

    public static final class QuestIds {
        public static final String FirstHunt = "first-hunt";
        public static final String OpenAuction = "open-auction";
        public static final String HerbGathering = "herb-gathering";

        private QuestIds() {}
    }

    public static final class QuestStatuses {
        public static final String InProgress = "InProgress";
        public static final String RewardGranted = "RewardGranted";

        private QuestStatuses() {}
    }

    public record KillMonsterReq(
            String playerId, String monsterId, String areaId, String idempotencyKey) {}

    public record KillMonsterRes(String eventId) {}

    public record CollectItemMsg(
            String playerId, String itemId, int count, String idempotencyKey) {}

    public record CompleteMissionReq(String playerId, String missionId, String idempotencyKey) {}

    public record CompleteMissionRes(String eventId) {}

    public record EnterAreaMsg(String playerId, String areaId, String idempotencyKey) {}

    public record UnlockFeatureReq(String playerId, String featureId, String idempotencyKey) {}

    public record UnlockFeatureRes(String eventId) {}

    public record JoinSessionReq(String playerId) {}

    public record JoinSessionRes(List<QuestProgress> activeQuests) {}

    public record GetQuestProgressReq(String playerId) {}

    public record GetQuestProgressRes(List<QuestProgress> activeQuests) {}

    public record DeleteQuestProjectionReq(String playerId, String questId) {}

    public record DeleteQuestProjectionRes(boolean deleted) {}

    public record RebuildQuestProjectionReq(String playerId, String questId, int count) {}

    public record SyncQuestProgressReq(String playerId) {}

    public record SyncQuestProgressRes(List<QuestProgress> updatedQuests) {}

    public record GetGameplaySnapshotReq(String playerId) {}

    public record GetGameplaySnapshotRes(
            String playerId,
            List<KillCountSnapshot> killCounts,
            List<ItemCountSnapshot> itemCounts,
            List<String> completedMissionIds,
            List<String> unlockedFeatureIds,
            List<String> enteredAreaIds,
            long snapshotVersion) {}

    public record KillCountSnapshot(String monsterId, String areaId, int count) {}

    public record ItemCountSnapshot(String itemId, int count) {}

    public record QuestProgressNotify(
            String playerId, String targetConnectionId, QuestProgress progress) {}

    public record QuestCompletedNotify(
            String playerId,
            String targetConnectionId,
            QuestProgress progress,
            boolean rewardGranted) {}

    public record QuestProgress(
            String playerId,
            String questId,
            String status,
            int currentCount,
            int requiredCount,
            String lastEventId,
            long updatedAtUnixMs) {}

    public record GameplayMsg(
            String eventId, String playerId, String type, byte[] payload, long occurredAtUnixMs) {
        public static GameplayMsg create(
                String eventId,
                String playerId,
                String type,
                String idempotencyKey,
                String value,
                int count,
                String sourceApi,
                long occurredAtUnixMs,
                boolean publish) {
            return new GameplayMsg(
                    eventId,
                    playerId,
                    type,
                    GameplayPayload.encode(
                            new GameplayPayload(idempotencyKey, value, count, sourceApi, publish)),
                    occurredAtUnixMs);
        }

        public GameplayPayload decodePayload() {
            return GameplayPayload.decode(payload);
        }

        public String idempotencyKey() {
            return decodePayload().idempotencyKey();
        }

        public String eventType() {
            return type;
        }

        public String value() {
            return decodePayload().value();
        }

        public int count() {
            return decodePayload().count();
        }

        public String sourceApi() {
            return decodePayload().sourceApi();
        }

        public long createdAtUnixMs() {
            return occurredAtUnixMs;
        }

        public boolean publish() {
            return decodePayload().publish();
        }
    }

    public record GameplayPayload(
            String idempotencyKey, String value, int count, String sourceApi, boolean publish) {
        private static byte[] encode(GameplayPayload payload) {
            try {
                ByteArrayOutputStream bytes = new ByteArrayOutputStream();
                try (DataOutputStream output = new DataOutputStream(bytes)) {
                    output.writeUTF(payload.idempotencyKey());
                    output.writeUTF(payload.value());
                    output.writeInt(payload.count());
                    output.writeUTF(payload.sourceApi());
                    output.writeBoolean(payload.publish());
                }
                return bytes.toByteArray();
            } catch (IOException error) {
                throw new IllegalStateException("failed to encode gameplay payload", error);
            }
        }

        private static GameplayPayload decode(byte[] bytes) {
            try (DataInputStream input = new DataInputStream(new ByteArrayInputStream(bytes))) {
                return new GameplayPayload(
                        input.readUTF(),
                        input.readUTF(),
                        input.readInt(),
                        input.readUTF(),
                        input.readBoolean());
            } catch (IOException error) {
                throw new IllegalArgumentException("invalid gameplay payload", error);
            }
        }
    }

    public record QuestProgressedEvent(
            String eventId,
            String playerId,
            String questId,
            int delta,
            int currentCount,
            int requiredCount,
            String sourceEventId) {}

    public record QuestCompletedEvent(
            String eventId,
            String playerId,
            String questId,
            String sourceEventId,
            long completedAtUnixMs) {}

    public record QuestRewardGrantedEvent(
            String eventId,
            String playerId,
            String questId,
            String sourceEventId,
            String rewardId,
            long grantedAtUnixMs) {}

    public record QuestProgressReconciledEvent(
            String eventId,
            String playerId,
            String questId,
            int currentCount,
            String reason,
            long reconciledAtUnixMs) {}

    public record StoredQuestEvent(
            String eventId,
            String sourceEventId,
            String playerId,
            String questId,
            String eventType,
            int delta,
            int currentCount,
            int requiredCount,
            String status,
            long version,
            long createdAtUnixMs) {}

    public record QuestProcessingMsg(
            String eventId,
            String playerId,
            List<QuestProgress> projection,
            List<QuestProgressNotify> progressNotifications,
            List<QuestCompletedNotify> completedNotifications,
            boolean duplicate) {}

    public record GameQuestServerAssertRes(boolean passed, List<String> evidence) {}
}

File: samples/kotlin/GameQuest/Shared/src/main/kotlin/systems/zlink/samples/kotlin/gamequest/shared/contracts/Messages.kt
package systems.zlink.samples.kotlin.gamequest.shared.contracts

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream

object QuestIds {
    const val FirstHunt = "first-hunt"
    const val OpenAuction = "open-auction"
    const val HerbGathering = "herb-gathering"
}

object QuestStatuses {
    const val InProgress = "InProgress"
    const val RewardGranted = "RewardGranted"
}

data class KillMonsterReq(
    val playerId: String,
    val monsterId: String,
    val areaId: String?,
    val idempotencyKey: String,
)

data class KillMonsterRes(val eventId: String)

data class CollectItemMsg(
    val playerId: String,
    val itemId: String,
    val count: Int,
    val idempotencyKey: String,
)

data class CompleteMissionReq(
    val playerId: String,
    val missionId: String,
    val idempotencyKey: String,
)

data class CompleteMissionRes(val eventId: String)

data class EnterAreaMsg(val playerId: String, val areaId: String, val idempotencyKey: String)

data class UnlockFeatureReq(
    val playerId: String,
    val featureId: String,
    val idempotencyKey: String,
)

data class UnlockFeatureRes(val eventId: String)

data class JoinSessionReq(val playerId: String)

data class JoinSessionRes(val activeQuests: List<QuestProgress>)

data class GetQuestProgressReq(val playerId: String)

data class GetQuestProgressRes(val activeQuests: List<QuestProgress>)

data class DeleteQuestProjectionReq(val playerId: String, val questId: String)

data class DeleteQuestProjectionRes(val deleted: Boolean)

data class RebuildQuestProjectionReq(val playerId: String, val questId: String, val count: Int)

data class SyncQuestProgressReq(val playerId: String)

data class SyncQuestProgressRes(val updatedQuests: List<QuestProgress>)

data class GetGameplaySnapshotReq(val playerId: String)

data class GetGameplaySnapshotRes(
    val playerId: String,
    val killCounts: List<KillCountSnapshot>,
    val itemCounts: List<ItemCountSnapshot>,
    val completedMissionIds: List<String>,
    val unlockedFeatureIds: List<String>,
    val enteredAreaIds: List<String>,
    val snapshotVersion: Long,
)

data class KillCountSnapshot(val monsterId: String, val areaId: String?, val count: Int)

data class ItemCountSnapshot(val itemId: String, val count: Int)

data class QuestProgressNotify(
    val playerId: String,
    val targetConnectionId: String?,
    val progress: QuestProgress,
)

data class QuestCompletedNotify(
    val playerId: String,
    val targetConnectionId: String?,
    val progress: QuestProgress,
    val rewardGranted: Boolean,
)

data class QuestProgress(
    val playerId: String,
    val questId: String,
    val status: String,
    val currentCount: Int,
    val requiredCount: Int,
    val lastEventId: String?,
    val updatedAtUnixMs: Long,
)

data class GameplayMsg(
    val eventId: String,
    val playerId: String,
    val type: String,
    val payload: ByteArray,
    val occurredAtUnixMs: Long,
) {
    fun decodePayload(): GameplayPayload = GameplayPayload.decode(payload)

    companion object {
        fun create(
            eventId: String,
            playerId: String,
            type: String,
            idempotencyKey: String,
            value: String,
            count: Int,
            sourceApi: String,
            occurredAtUnixMs: Long,
            publish: Boolean,
        ): GameplayMsg =
            GameplayMsg(
                eventId,
                playerId,
                type,
                GameplayPayload.encode(
                    GameplayPayload(idempotencyKey, value, count, sourceApi, publish)
                ),
                occurredAtUnixMs,
            )
    }
}

data class GameplayPayload(
    val idempotencyKey: String,
    val value: String,
    val count: Int,
    val sourceApi: String,
    val publish: Boolean,
) {
    companion object {
        fun encode(payload: GameplayPayload): ByteArray {
            val bytes = ByteArrayOutputStream()
            DataOutputStream(bytes).use { output ->
                output.writeUTF(payload.idempotencyKey)
                output.writeUTF(payload.value)
                output.writeInt(payload.count)
                output.writeUTF(payload.sourceApi)
                output.writeBoolean(payload.publish)
            }
            return bytes.toByteArray()
        }

        fun decode(bytes: ByteArray): GameplayPayload =
            DataInputStream(ByteArrayInputStream(bytes)).use { input ->
                GameplayPayload(
                    input.readUTF(),
                    input.readUTF(),
                    input.readInt(),
                    input.readUTF(),
                    input.readBoolean(),
                )
            }
    }
}

class QuestProgressedEvent

class QuestCompletedEvent

class QuestRewardGrantedEvent

class QuestProgressReconciledEvent

data class StoredQuestEvent(
    val eventId: String,
    val sourceEventId: String?,
    val playerId: String,
    val questId: String,
    val eventType: String,
    val currentCount: Int,
    val requiredCount: Int,
    val status: String,
    val version: Long,
    val createdAtUnixMs: Long,
)

data class QuestProcessingMsg(
    val eventId: String,
    val playerId: String,
    val projection: List<QuestProgress>,
    val progressNotifications: List<QuestProgressNotify>,
    val completedNotifications: List<QuestCompletedNotify>,
    val duplicate: Boolean,
)

data class GameQuestServerAssertRes(val passed: Boolean, val evidence: List<String>)
